### ISSUE TITLE
CA1031 should thread `return Task.FromException(...)` as `throw`

### ISSUE THREAD

**Issue by Pretasoc** [dotnet/roslyn-analyzers#3748]
#### Analyzer package

[Microsoft.CodeQuality.Analyzers](https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers/)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

Example: [CA1031(https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1031?view=vs-2019)

#### Repro steps

Consider the following snippet:

``` csharp
public Task FooAsync()
{
    try
    {
        DoStuff()
        return Task.CompletedTask;
    }
    catch(Exception ex)
    {
        return Task.FromException(ex);
    }

}
```

#### Expected behavior

The code compiles without a warning. In my understanding returning a failed task from a method, with an awaitable result is the equivalent on rethrowing an exception in a non awaitable method.

#### Actual behavior

A `CA1031` is raised on the `catch(Exception ex)` line.

**Comment by jinek** [comment]
In general a stack can not be recovered by simple copying of the exception reference.

This is the idea of CA1031: our code should stay transparent for unexpected exceptions, thus allowing the code which "is expecting" to do the job.

For example, exceptions thrown in DoStuff will not stop the debugger anymore (no more possible to see the stack, evaluate debugging stuff etc).

Another example, is if somebody calls FooAsync(); without of even checking the returned value. In this case all exception there are just lost.
